In [1]:
# GroupBy Attributes and Methods
# find total number of groups -> LEN AND NUNIQUE
# find items in each group -> size
# first()/last() -> nth item
# get_group -> vs filtering
# groups
# describe
# sample
# nunique
# split apply and combine strategy with functions

In [2]:
import pandas as pd
import numpy as np

In [3]:
from google.colab import drive
drive.mount('Python')

Mounted at Python


In [4]:
# 1. group by
# 2.

In [5]:
movies = pd.read_csv('/content/Python/MyDrive/Colab Notebooks/Colab_Files_Data/Python/DataFrames/imdb-top-1000.csv')

In [6]:
movies.head()

Series_Title Released_Year  Runtime   Genre  IMDB_Rating  \
0  The Shawshank Redemption          1994      142   Drama          9.3   
1             The Godfather          1972      175   Crime          9.2   
2           The Dark Knight          2008      152  Action          9.0   
3    The Godfather: Part II          1974      202   Crime          9.0   
4              12 Angry Men          1957       96   Crime          9.0   

               Director           Star1  No_of_Votes        Gross  Metascore  
0        Frank Darabont     Tim Robbins      2343110   28341469.0       80.0  
1  Francis Ford Coppola   Marlon Brando      1620367  134966411.0      100.0  
2     Christopher Nolan  Christian Bale      2303232  534858444.0       84.0  
3  Francis Ford Coppola       Al Pacino      1129952   57300000.0       90.0  
4          Sidney Lumet     Henry Fonda       689845    4360000.0       96.0

In [7]:
genres = movies.groupby('Genre')   ############################### Returns group by object ####################

In [8]:
genres.min()

Series_Title Released_Year  Runtime  \
Genre                                                               
Action                                 300          1924       45   
Adventure            2001: A Space Odyssey          1925       88   
Animation                            Akira          1940       71   
Biography                 12 Years a Slave          1928       93   
Comedy                (500) Days of Summer          1921       68   
Crime                         12 Angry Men          1931       80   
Drama                                 1917          1925       64   
Family          E.T. the Extra-Terrestrial          1971      100   
Fantasy       Das Cabinet des Dr. Caligari          1920       76   
Film-Noir                Shadow of a Doubt          1941      100   
Horror                               Alien          1933       71   
Mystery                          Dark City          1938       96   
Thriller                   Wait Until Dark          1967      108   
Western    Il buono, il brutto, il cattivo          1965      132   

           IMDB_Rating               Director                     Star1  \
Genre                                                                     
Action             7.6       Abhishek Chaubey                Aamir Khan   
Adventure          7.6         Akira Kurosawa                Aamir Khan   
Animation          7.6            Adam Elliot             Adrian Molina   
Biography          7.6             Adam McKay              Adrien Brody   
Comedy             7.6  Alejandro G. Iñárritu                Aamir Khan   
Crime              7.6         Akira Kurosawa                Ajay Devgn   
Drama              7.6             Aamir Khan                Abhay Deol   
Family             7.8             Mel Stuart               Gene Wilder   
Fantasy            7.9            F.W. Murnau               Max Schreck   
Film-Noir          7.8       Alfred Hitchcock           Humphrey Bogart   
Horror             7.6     Alejandro Amenábar           Anthony Perkins   
Mystery            7.6            Alex Proyas  Bernard-Pierre Donnadieu   
Thriller           7.8          Terence Young            Audrey Hepburn   
Western            7.8         Clint Eastwood            Clint Eastwood   

           No_of_Votes        Gross  Metascore  
Genre                                           
Action           25312       3296.0       33.0  
Adventure        29999      61001.0       41.0  
Animation        25229     128985.0       61.0  
Biography        27254      21877.0       48.0  
Comedy           26337       1305.0       45.0  
Crime            27712       6013.0       47.0  
Drama            25088       3600.0       28.0  
Family          178731    4000000.0       67.0  
Fantasy          57428  337574718.0        NaN  
Film-Noir        59556     449191.0       94.0  
Horror           27007      89029.0       46.0  
Mystery          33982    1035953.0       52.0  
Thriller         27733   17550741.0       81.0  
Western          65659    5321508.0       69.0

In [9]:
# Applying builtin aggregation fuctions on groupby objects
# genres.std()

In [10]:
# find the top 3 genres by total earning
movies.groupby('Genre').sum()['Gross'].sort_values(ascending=False).head(3)

Genre
Drama     3.540997e+10
Action    3.263226e+10
Comedy    1.566387e+10
Name: Gross, dtype: float64

In [11]:
# find the genre with highest avg IMDB rating
movies.groupby('Genre')['IMDB_Rating'].mean().sort_values(ascending=False).head(1)

Genre
Western    8.35
Name: IMDB_Rating, dtype: float64

In [12]:
# find director with most popularity
movies.groupby('Director')['No_of_Votes'].sum().sort_values(ascending=False).head(1)

Director
Christopher Nolan    11578345
Name: No_of_Votes, dtype: int64

In [13]:
# find the highest rated movie of each genre
movies.groupby('Genre')['IMDB_Rating'].max()

Genre
Action       9.0
Adventure    8.6
Animation    8.6
Biography    8.9
Comedy       8.6
Crime        9.2
Drama        9.3
Family       7.8
Fantasy      8.1
Film-Noir    8.1
Horror       8.5
Mystery      8.4
Thriller     7.8
Western      8.8
Name: IMDB_Rating, dtype: float64

In [14]:
# find number of movies done by each actor
# movies['Star1'].value_counts()
movies.groupby('Star1')['Series_Title'].count().sort_values(ascending=False)

Star1
Tom Hanks             12
Robert De Niro        11
Clint Eastwood        10
Al Pacino             10
Leonardo DiCaprio      9
                      ..
Glen Hansard           1
Giuseppe Battiston     1
Giulietta Masina       1
Gerardo Taracena       1
Ömer Faruk Sorak       1
Name: Series_Title, Length: 660, dtype: int64

In [15]:
# GroupBy Attributes and Methods
# find total number of groups -> LEN AND NUNIQUE
# find items in each group -> size
# first()/last() -> nth item
# get_group -> vs filtering
# groups                    ########################
# groups.keys() return the group names in list as keys of the dictionary of groups   ############
# describe
# sample
# nunique

In [16]:
len(movies.groupby('Genre'))

14

In [17]:
movies['Genre'].nunique()

14

In [18]:
movies.groupby('Genre').nunique()        # Any function we apply gets applied on all columns within each group

Series_Title  Released_Year  Runtime  IMDB_Rating  Director  Star1  \
Genre                                                                           
Action              172             61       78           15       123    121   
Adventure            72             49       58           10        59     59   
Animation            82             35       41           11        51     77   
Biography            88             44       56           13        76     72   
Comedy              155             72       70           11       113    133   
Crime               106             56       65           14        86     85   
Drama               289             83       95           14       211    250   
Family                2              2        2            1         2      2   
Fantasy               2              2        2            2         2      2   
Film-Noir             3              3        3            3         3      3   
Horror               11             11       10            8        10     11   
Mystery              12             11       10            8        10     11   
Thriller              1              1        1            1         1      1   
Western               4              4        4            4         2      2   

           No_of_Votes  Gross  Metascore  
Genre                                     
Action             172    172         50  
Adventure           72     72         33  
Animation           82     82         29  
Biography           88     88         40  
Comedy             155    155         44  
Crime              107    107         39  
Drama              288    287         52  
Family               2      2          2  
Fantasy              2      2          0  
Film-Noir            3      3          3  
Horror              11     11          9  
Mystery             12     12          7  
Thriller             1      1          1  
Western              4      4          4

In [19]:
g=movies.groupby('Genre')

In [20]:
genres = movies.groupby('Genre')
#genres.first()                                 ######## First movie of each genre
# genres.last()
genres.nth(15)                                ########## (15th movies of each genre)

Series_Title Released_Year  Runtime      Genre  \
59                   Avengers: Endgame          2019      181     Action   
67                 The Lives of Others          2006      137      Drama   
87                            Drishyam          2013      160      Crime   
153                    Rang De Basanti          2006      167     Comedy   
205                               Soul          2020      100  Animation   
212                      Hacksaw Ridge          2016      139  Biography   
247  Lagaan: Once Upon a Time in India          2001      224  Adventure   

     IMDB_Rating                          Director            Star1  \
59           8.4                     Anthony Russo        Joe Russo   
67           8.4  Florian Henckel von Donnersmarck      Ulrich Mühe   
87           8.3                     Jeethu Joseph         Mohanlal   
153          8.2           Rakeysh Omprakash Mehra       Aamir Khan   
205          8.1                       Pete Docter      Kemp Powers   
212          8.1                        Mel Gibson  Andrew Garfield   
247          8.1                Ashutosh Gowariker       Aamir Khan   

     No_of_Votes        Gross  Metascore  
59        809955  858373000.0       78.0  
67        358685   11286112.0       89.0  
87         30722   11339376.0        NaN  
153       111937    2197331.0        NaN  
205       159171  321826155.0       83.0  
212       435928   67209615.0       71.0  
247       105036      70147.0       84.0

In [21]:
genres.nth(15)

Series_Title Released_Year  Runtime      Genre  \
59                   Avengers: Endgame          2019      181     Action   
67                 The Lives of Others          2006      137      Drama   
87                            Drishyam          2013      160      Crime   
153                    Rang De Basanti          2006      167     Comedy   
205                               Soul          2020      100  Animation   
212                      Hacksaw Ridge          2016      139  Biography   
247  Lagaan: Once Upon a Time in India          2001      224  Adventure   

     IMDB_Rating                          Director            Star1  \
59           8.4                     Anthony Russo        Joe Russo   
67           8.4  Florian Henckel von Donnersmarck      Ulrich Mühe   
87           8.3                     Jeethu Joseph         Mohanlal   
153          8.2           Rakeysh Omprakash Mehra       Aamir Khan   
205          8.1                       Pete Docter      Kemp Powers   
212          8.1                        Mel Gibson  Andrew Garfield   
247          8.1                Ashutosh Gowariker       Aamir Khan   

     No_of_Votes        Gross  Metascore  
59        809955  858373000.0       78.0  
67        358685   11286112.0       89.0  
87         30722   11339376.0        NaN  
153       111937    2197331.0        NaN  
205       159171  321826155.0       83.0  
212       435928   67209615.0       71.0  
247       105036      70147.0       84.0

In [22]:
movies['Genre'].value_counts()

Genre
Drama        289
Action       172
Comedy       155
Crime        107
Biography     88
Animation     82
Adventure     72
Mystery       12
Horror        11
Western        4
Film-Noir      3
Fantasy        2
Family         2
Thriller       1
Name: count, dtype: int64

In [23]:
genres.get_group('Drama')    ################### to get any genre ##############

Series_Title Released_Year  Runtime  Genre  \
0           The Shawshank Redemption          1994      142  Drama   
9                         Fight Club          1999      139  Drama   
11                      Forrest Gump          1994      142  Drama   
17   One Flew Over the Cuckoo's Nest          1975      133  Drama   
20                   Soorarai Pottru          2020      153  Drama   
..                               ...           ...      ...    ...   
990                     Giù la testa          1971      157  Drama   
993                           Blowup          1966      111  Drama   
996                            Giant          1956      201  Drama   
997            From Here to Eternity          1953      118  Drama   
998                         Lifeboat          1944       97  Drama   

     IMDB_Rating                Director              Star1  No_of_Votes  \
0            9.3          Frank Darabont        Tim Robbins      2343110   
9            8.8           David Fincher          Brad Pitt      1854740   
11           8.8         Robert Zemeckis          Tom Hanks      1809221   
17           8.7            Milos Forman     Jack Nicholson       918088   
20           8.6           Sudha Kongara             Suriya        54995   
..           ...                     ...                ...          ...   
990          7.6            Sergio Leone        Rod Steiger        30144   
993          7.6  Michelangelo Antonioni     David Hemmings        56513   
996          7.6          George Stevens   Elizabeth Taylor        34075   
997          7.6          Fred Zinnemann     Burt Lancaster        43374   
998          7.6        Alfred Hitchcock  Tallulah Bankhead        26471   

           Gross  Metascore  
0     28341469.0       80.0  
9     37030102.0       66.0  
11   330252182.0       82.0  
17   112000000.0       83.0  
20   556832648.0        NaN  
..           ...        ...  
990     696690.0       77.0  
993  632532802.0       82.0  
996  195217415.0       84.0  
997   30500000.0       85.0  
998  852142728.0       78.0  

[289 rows x 10 columns]

In [24]:
genres.get_group('Fantasy')
movies[movies['Genre']=='Fantasy']

Series_Title Released_Year  Runtime    Genre  \
321  Das Cabinet des Dr. Caligari          1920       76  Fantasy   
568                     Nosferatu          1922       94  Fantasy   

     IMDB_Rating      Director          Star1  No_of_Votes        Gross  \
321          8.1  Robert Wiene  Werner Krauss        57428  337574718.0   
568          7.9   F.W. Murnau    Max Schreck        88794  445151978.0   

     Metascore  
321        NaN  
568        NaN

In [25]:
genres.groups  ################ Index position of each group items ###### (sort of dictionary)

{'Action': [2, 5, 8, 10, 13, 14, 16, 29, 30, 31, 39, 42, 44, 55, 57, 59, 60, 63, 68, 72, 106, 109, 129, 130, 134, 140, 142, 144, 152, 155, 160, 161, 166, 168, 171, 172, 177, 181, 194, 201, 202, 216, 217, 223, 224, 236, 241, 262, 275, 294, 308, 320, 325, 326, 331, 337, 339, 340, 343, 345, 348, 351, 353, 356, 357, 362, 368, 369, 375, 376, 390, 410, 431, 436, 473, 477, 479, 482, 488, 493, 496, 502, 507, 511, 532, 535, 540, 543, 564, 569, 570, 573, 577, 582, 583, 602, 605, 608, 615, 623, ...], 'Adventure': [21, 47, 93, 110, 114, 116, 118, 137, 178, 179, 191, 193, 209, 226, 231, 247, 267, 273, 281, 300, 301, 304, 306, 323, 329, 361, 366, 377, 402, 406, 415, 426, 458, 470, 497, 498, 506, 513, 514, 537, 549, 552, 553, 566, 576, 604, 609, 618, 638, 647, 675, 681, 686, 692, 711, 713, 739, 755, 781, 797, 798, 851, 873, 884, 912, 919, 947, 957, 964, 966, 984, 991], 'Animation': [23, 43, 46, 56, 58, 61, 66, 70, 101, 135, 146, 151, 158, 170, 197, 205, 211, 213, 219, 229, 230, 242, 245, 246, 270, 33

In [26]:
genres.describe()

Runtime                                                              \
            count        mean        std    min     25%    50%     75%    max   
Genre                                                                           
Action      172.0  129.046512  28.500706   45.0  110.75  127.5  143.25  321.0   
Adventure    72.0  134.111111  33.317320   88.0  109.00  127.0  149.00  228.0   
Animation    82.0   99.585366  14.530471   71.0   90.00   99.5  106.75  137.0   
Biography    88.0  136.022727  25.514466   93.0  120.00  129.0  146.25  209.0   
Comedy      155.0  112.129032  22.946213   68.0   96.00  106.0  124.50  188.0   
Crime       107.0  126.392523  27.689231   80.0  106.50  122.0  141.50  229.0   
Drama       289.0  124.737024  27.740490   64.0  105.00  121.0  137.00  242.0   
Family        2.0  107.500000  10.606602  100.0  103.75  107.5  111.25  115.0   
Fantasy       2.0   85.000000  12.727922   76.0   80.50   85.0   89.50   94.0   
Film-Noir     3.0  104.000000   4.000000  100.0  102.00  104.0  106.00  108.0   
Horror       11.0  102.090909  13.604812   71.0   98.00  103.0  109.00  122.0   
Mystery      12.0  119.083333  14.475423   96.0  110.75  117.5  130.25  138.0   
Thriller      1.0  108.000000        NaN  108.0  108.00  108.0  108.00  108.0   
Western       4.0  148.250000  17.153717  132.0  134.25  148.0  162.00  165.0   

          IMDB_Rating            ...         Gross              Metascore  \
                count      mean  ...           75%          max     count   
Genre                            ...                                        
Action          172.0  7.949419  ...  2.674437e+08  936662225.0     143.0   
Adventure        72.0  7.937500  ...  1.998070e+08  874211619.0      64.0   
Animation        82.0  7.930488  ...  2.520612e+08  873839108.0      75.0   
Biography        88.0  7.938636  ...  9.829924e+07  753585104.0      79.0   
Comedy          155.0  7.901290  ...  8.107809e+07  886752933.0     125.0   
Crime           107.0  8.016822  ...  7.102163e+07  790482117.0      87.0   
Drama           289.0  7.957439  ...  1.164461e+08  924558264.0     241.0   
Family            2.0  7.800000  ...  3.273329e+08  435110554.0       2.0   
Fantasy           2.0  8.000000  ...  4.182577e+08  445151978.0       0.0   
Film-Noir         3.0  7.966667  ...  6.273068e+07  123353292.0       3.0   
Horror           11.0  7.909091  ...  1.362817e+08  298791505.0      11.0   
Mystery          12.0  7.975000  ...  1.310949e+08  474203697.0       8.0   
Thriller          1.0  7.800000  ...  1.755074e+07   17550741.0       1.0   
Western           4.0  8.350000  ...  1.920000e+07   31800000.0       4.0   

                                                                  
                mean        std   min    25%   50%    75%    max  
Genre                                                             
Action     73.419580  12.421252  33.0  65.00  74.0  82.00   98.0  
Adventure  78.437500  12.345393  41.0  69.75  80.5  87.25  100.0  
Animation  81.093333   8.813646  61.0  75.00  82.0  87.50   96.0  
Biography  76.240506  11.028187  48.0  70.50  76.0  84.50   97.0  
Comedy     78.720000  11.829160  45.0  72.00  79.0  88.00   99.0  
Crime      77.080460  13.099102  47.0  69.50  77.0  87.00  100.0  
Drama      79.701245  12.744687  28.0  72.00  82.0  89.00  100.0  
Family     79.000000  16.970563  67.0  73.00  79.0  85.00   91.0  
Fantasy          NaN        NaN   NaN    NaN   NaN    NaN    NaN  
Film-Noir  95.666667   1.527525  94.0  95.00  96.0  96.50   97.0  
Horror     80.000000  15.362291  46.0  77.50  87.0  88.50   97.0  
Mystery    79.125000  18.604435  52.0  65.25  77.0  98.50  100.0  
Thriller   81.000000        NaN  81.0  81.00  81.0  81.00   81.0  
Western    78.250000   9.032349  69.0  72.75  77.0  82.50   90.0  

[14 rows x 40 columns]

In [27]:
genres.sample(2,replace=True)      ############# random 2 movies from each genres and replace for handing the case where
#any genre having only one movie so it repeats that movie###########

Series_Title Released_Year  Runtime      Genre  \
390            Knockin' on Heaven's Door          1997       87     Action   
853                          Dirty Harry          1971      102     Action   
137                                Queen          2013      146  Adventure   
178      Monty Python and the Holy Grail          1975       91  Adventure   
771                              Tangled          2010      100  Animation   
395       The Nightmare Before Christmas          1993       76  Animation   
131                           Green Book          2018      130  Biography   
949                                 Blow          2001      124  Biography   
972                         Delicatessen          1991       99     Comedy   
445                            Stalag 17          1953      120     Comedy   
239                       Salinui chueok          2003      131      Crime   
37                          The Departed          2006      151      Crime   
580                Manchester by the Sea          2016      137      Drama   
206                           Kis Uykusu          2014      196      Drama   
698  Willy Wonka & the Chocolate Factory          1971      100     Family   
698  Willy Wonka & the Chocolate Factory          1971      100     Family   
321         Das Cabinet des Dr. Caligari          1920       76    Fantasy   
568                            Nosferatu          1922       94    Fantasy   
309                        The Third Man          1949      104  Film-Noir   
456                   The Maltese Falcon          1941      100  Film-Noir   
49                                Psycho          1960      109     Horror   
707                        The Innocents          1961      100     Horror   
899                            El cuerpo          2012      112    Mystery   
145                       Shutter Island          2010      138    Mystery   
700                      Wait Until Dark          1967      108   Thriller   
700                      Wait Until Dark          1967      108   Thriller   
691               The Outlaw Josey Wales          1976      135    Western   
12       Il buono, il brutto, il cattivo          1966      161    Western   

     IMDB_Rating           Director               Star1  No_of_Votes  \
390          8.0        Thomas Jahn       Til Schweiger        27721   
853          7.7         Don Siegel      Clint Eastwood       143292   
137          8.2         Vikas Bahl      Kangana Ranaut        60701   
178          8.2      Terry Gilliam         Terry Jones       500875   
771          7.7       Nathan Greno        Byron Howard       405922   
395          8.0       Henry Selick        Danny Elfman       300208   
131          8.2     Peter Farrelly     Viggo Mortensen       377884   
949          7.6          Ted Demme         Johnny Depp       240714   
972          7.6          Marc Caro  Jean-Pierre Jeunet        80487   
445          8.0       Billy Wilder      William Holden        51046   
239          8.1       Bong Joon Ho        Kang-ho Song       139558   
37           8.5    Martin Scorsese   Leonardo DiCaprio      1189773   
580          7.8   Kenneth Lonergan       Casey Affleck       246963   
206          8.1  Nuri Bilge Ceylan      Haluk Bilginer        46547   
698          7.8         Mel Stuart         Gene Wilder       178731   
698          7.8         Mel Stuart         Gene Wilder       178731   
321          8.1       Robert Wiene       Werner Krauss        57428   
568          7.9        F.W. Murnau         Max Schreck        88794   
309          8.1         Carol Reed        Orson Welles       158731   
456          8.0        John Huston     Humphrey Bogart       148928   
49           8.5   Alfred Hitchcock     Anthony Perkins       604211   
707          7.8       Jack Clayton        Deborah Kerr        27007   
899          7.6        Oriol Paulo       Jose Coronado        57549   
145          8.2    Martin Scorsese  

In [28]:
genres.nunique()

Series_Title  Released_Year  Runtime  IMDB_Rating  Director  Star1  \
Genre                                                                           
Action              172             61       78           15       123    121   
Adventure            72             49       58           10        59     59   
Animation            82             35       41           11        51     77   
Biography            88             44       56           13        76     72   
Comedy              155             72       70           11       113    133   
Crime               106             56       65           14        86     85   
Drama               289             83       95           14       211    250   
Family                2              2        2            1         2      2   
Fantasy               2              2        2            2         2      2   
Film-Noir             3              3        3            3         3      3   
Horror               11             11       10            8        10     11   
Mystery              12             11       10            8        10     11   
Thriller              1              1        1            1         1      1   
Western               4              4        4            4         2      2   

           No_of_Votes  Gross  Metascore  
Genre                                     
Action             172    172         50  
Adventure           72     72         33  
Animation           82     82         29  
Biography           88     88         40  
Comedy             155    155         44  
Crime              107    107         39  
Drama              288    287         52  
Family               2      2          2  
Fantasy              2      2          0  
Film-Noir            3      3          3  
Horror              11     11          9  
Mystery             12     12          7  
Thriller             1      1          1  
Western              4      4          4

In [29]:
# genres.agg(['mean','max']) # doesn't work like this

In [30]:
genres.agg(
    {'Runtime':['mean','max'],
    'IMDB_Rating':'mean'
    }
)

Runtime      IMDB_Rating
                 mean  max        mean
Genre                                 
Action     129.046512  321    7.949419
Adventure  134.111111  228    7.937500
Animation   99.585366  137    7.930488
Biography  136.022727  209    7.938636
Comedy     112.129032  188    7.901290
Crime      126.392523  229    8.016822
Drama      124.737024  242    7.957439
Family     107.500000  115    7.800000
Fantasy     85.000000   94    8.000000
Film-Noir  104.000000  108    7.966667
Horror     102.090909  122    7.909091
Mystery    119.083333  138    7.975000
Thriller   108.000000  108    7.800000
Western    148.250000  165    8.350000

In [31]:
# agg method                        #########################################################
# passing dict
genres.agg(
    {
        'Runtime':'mean',
        'IMDB_Rating':'mean',
        'No_of_Votes':'sum',
        'Gross':'sum',
        'Metascore':'min'
    }
)

Runtime  IMDB_Rating  No_of_Votes         Gross  Metascore
Genre                                                                   
Action     129.046512     7.949419     72282412  3.263226e+10       33.0
Adventure  134.111111     7.937500     22576163  9.496922e+09       41.0
Animation   99.585366     7.930488     21978630  1.463147e+10       61.0
Biography  136.022727     7.938636     24006844  8.276358e+09       48.0
Comedy     112.129032     7.901290     27620327  1.566387e+10       45.0
Crime      126.392523     8.016822     33533615  8.452632e+09       47.0
Drama      124.737024     7.957439     61367304  3.540997e+10       28.0
Family     107.500000     7.800000       551221  4.391106e+08       67.0
Fantasy     85.000000     8.000000       146222  7.827267e+08        NaN
Film-Noir  104.000000     7.966667       367215  1.259105e+08       94.0
Horror     102.090909     7.909091      3742556  1.034649e+09       46.0
Mystery    119.083333     7.975000      4203004  1.256417e+09       52.0
Thriller   108.000000     7.800000        27733  1.755074e+07       81.0
Western    148.250000     8.350000      1289665  5.822151e+07       69.0

In [32]:
# passing list
# genres.agg(['min','max','mean','sum'])

In [33]:
genres.agg(                    ########## specific function on any column with the help of dictionary can even specify the multiple agg fucntion
    {
        'IMDB_Rating':['min','max','std'],
        'No_of_Votes':['sum','max'],                          ########Combining both the syntaxes ############
        'Gross':'sum'
    }
)

IMDB_Rating                No_of_Votes                  Gross
                  min  max       std         sum      max           sum
Genre                                                                  
Action            7.6  9.0  0.304258    72282412  2303232  3.263226e+10
Adventure         7.6  8.6  0.229781    22576163  1512360  9.496922e+09
Animation         7.6  8.6  0.253221    21978630   999790  1.463147e+10
Biography         7.6  8.9  0.267140    24006844  1213505  8.276358e+09
Comedy            7.6  8.6  0.228771    27620327   939631  1.566387e+10
Crime             7.6  9.2  0.335477    33533615  1826188  8.452632e+09
Drama             7.6  9.3  0.267229    61367304  2343110  3.540997e+10
Family            7.8  7.8  0.000000      551221   372490  4.391106e+08
Fantasy           7.9  8.1  0.141421      146222    88794  7.827267e+08
Film-Noir         7.8  8.1  0.152753      367215   158731  1.259105e+08
Horror            7.6  8.5  0.311302     3742556   787806  1.034649e+09
Mystery           7.6  8.4  0.310791     4203004  1129894  1.256417e+09
Thriller          7.8  7.8       NaN       27733    27733  1.755074e+07
Western           7.8  8.8  0.420317     1289665   688390  5.822151e+07

In [34]:
# looping on groups
# Two objects group ,data first ones is name of the group where as next one is dataframe of that group so you can treat
# second object as a dataframe and do the manipulation
df = pd.DataFrame(columns=movies.columns)
for group,data in genres:
  df = pd.concat((df, data[data['IMDB_Rating'] == data['IMDB_Rating'].max()]))

In [35]:
#  pd.concat            ################# important and to be used all the time
df=pd.DataFrame(columns=movies.columns)
for group,data in genres:
  max=data[data['IMDB_Rating']==data['IMDB_Rating'].max()]
  df=pd.concat((df,max))

In [36]:
# split (apply) combine
# apply -> builtin function
genres.apply(min)

Series_Title Released_Year  Runtime      Genre  \
Genre                                                                          
Action                                 300          1924       45     Action   
Adventure            2001: A Space Odyssey          1925       88  Adventure   
Animation                            Akira          1940       71  Animation   
Biography                 12 Years a Slave          1928       93  Biography   
Comedy                (500) Days of Summer          1921       68     Comedy   
Crime                         12 Angry Men          1931       80      Crime   
Drama                                 1917          1925       64      Drama   
Family          E.T. the Extra-Terrestrial          1971      100     Family   
Fantasy       Das Cabinet des Dr. Caligari          1920       76    Fantasy   
Film-Noir                Shadow of a Doubt          1941      100  Film-Noir   
Horror                               Alien          1933       71     Horror   
Mystery                          Dark City          1938       96    Mystery   
Thriller                   Wait Until Dark          1967      108   Thriller   
Western    Il buono, il brutto, il cattivo          1965      132    Western   

           IMDB_Rating               Director                     Star1  \
Genre                                                                     
Action             7.6       Abhishek Chaubey                Aamir Khan   
Adventure          7.6         Akira Kurosawa                Aamir Khan   
Animation          7.6            Adam Elliot             Adrian Molina   
Biography          7.6             Adam McKay              Adrien Brody   
Comedy             7.6  Alejandro G. Iñárritu                Aamir Khan   
Crime              7.6         Akira Kurosawa                Ajay Devgn   
Drama              7.6             Aamir Khan                Abhay Deol   
Family             7.8             Mel Stuart               Gene Wilder   
Fantasy            7.9            F.W. Murnau               Max Schreck   
Film-Noir          7.8       Alfred Hitchcock           Humphrey Bogart   
Horror             7.6     Alejandro Amenábar           Anthony Perkins   
Mystery            7.6            Alex Proyas  Bernard-Pierre Donnadieu   
Thriller           7.8          Terence Young            Audrey Hepburn   
Western            7.8         Clint Eastwood            Clint Eastwood   

           No_of_Votes        Gross  Metascore  
Genre                                           
Action           25312       3296.0        NaN  
Adventure        29999      61001.0        NaN  
Animation        25229     128985.0        NaN  
Biography        27254      21877.0        NaN  
Comedy           26337       1305.0        NaN  
Crime            27712       6013.0        NaN  
Drama            25088       3600.0        NaN  
Family          178731    4000000.0       67.0  
Fantasy          57428  337574718.0        NaN  
Film-Noir        59556     449191.0       94.0  
Horror           27007      89029.0       46.0  
Mystery          33982    1035953.0        NaN  
Thriller         27733   17550741.0       81.0  
Western          65659    5321508.0       69.0

In [37]:
# How many movies in each genre stars with 'A'
count=0
for group,data in genres:                   ####### How loops works are very important understand loop works for each group
  count=data['Series_Title'].str.startswith('A').sum()
  print(group,count)

Action 10
Adventure 2
Animation 2
Biography 9
Comedy 14
Crime 4
Drama 21
Family 0
Fantasy 0
Film-Noir 0
Horror 1
Mystery 0
Thriller 0
Western 0


In [38]:
genres=movies.groupby('Genre')
def foo(df):
  return df['Series_Title'].str.startswith('A').sum()
genres.apply(foo)

Genre
Action       10
Adventure     2
Animation     2
Biography     9
Comedy       14
Crime         4
Drama        21
Family        0
Fantasy       0
Film-Noir     0
Horror        1
Mystery       0
Thriller      0
Western       0
dtype: int64

In [39]:
import pandas as pd
df1 = pd.DataFrame(columns=movies.columns)
for group, data in genres:
    max_rating_row = data[data['IMDB_Rating'] == data['IMDB_Rating'].max()]
    df1 = pd.concat([df1, max_rating_row])
df1

Series_Title Released_Year Runtime      Genre  \
2                        The Dark Knight          2008     152     Action   
21                          Interstellar          2014     169  Adventure   
23         Sen to Chihiro no kamikakushi          2001     125  Animation   
7                       Schindler's List          1993     195  Biography   
19                          Gisaengchung          2019     132     Comedy   
26                       La vita è bella          1997     116     Comedy   
1                          The Godfather          1972     175      Crime   
0               The Shawshank Redemption          1994     142      Drama   
688           E.T. the Extra-Terrestrial          1982     115     Family   
698  Willy Wonka & the Chocolate Factory          1971     100     Family   
321         Das Cabinet des Dr. Caligari          1920      76    Fantasy   
309                        The Third Man          1949     104  Film-Noir   
49                                Psycho          1960     109     Horror   
69                               Memento          2000     113    Mystery   
81                           Rear Window          1954     112    Mystery   
700                      Wait Until Dark          1967     108   Thriller   
12       Il buono, il brutto, il cattivo          1966     161    Western   

     IMDB_Rating              Director                Star1 No_of_Votes  \
2            9.0     Christopher Nolan       Christian Bale     2303232   
21           8.6     Christopher Nolan  Matthew McConaughey     1512360   
23           8.6        Hayao Miyazaki        Daveigh Chase      651376   
7            8.9      Steven Spielberg          Liam Neeson     1213505   
19           8.6          Bong Joon Ho         Kang-ho Song      552778   
26           8.6       Roberto Benigni      Roberto Benigni      623629   
1            9.2  Francis Ford Coppola        Marlon Brando     1620367   
0            9.3        Frank Darabont          Tim Robbins     2343110   
688          7.8      Steven Spielberg         Henry Thomas      372490   
698          7.8            Mel Stuart          Gene Wilder      178731   
321          8.1          Robert Wiene        Werner Krauss       57428   
309          8.1            Carol Reed         Orson Welles      158731   
49           8.5      Alfred Hitchcock      Anthony Perkins      604211   
69           8.4     Christopher Nolan           Guy Pearce     1125712   
81           8.4      Alfred Hitchcock        James Stewart      444074   
700          7.8         Terence Young       Audrey Hepburn       27733   
12           8.8          Sergio Leone       Clint Eastwood      688390   

           Gross  Metascore  
2    534858444.0       84.0  
21   188020017.0       74.0  
23    10055859.0       96.0  
7     96898818.0       94.0  
19    53367844.0       96.0  
26    57598247.0       59.0  
1    134966411.0      100.0  
0     28341469.0       80.0  
688  435110554.0       91.0  
698    4000000.0       67.0  
321  337574718.0        NaN  
309     449191.0       97.0  
49    32000000.0       97.0  
69    25544867.0       80.0  
81    36764313.0      100.0  
700   17550741.0       81.0  
12     6100000.0       90.0

In [40]:
df1=pd.DataFrame(columns=movies.columns)
for group,data in genres:
  df1=pd.concat((df1,data[data['IMDB_Rating']==data['IMDB_Rating'].max()]))   ### assign back the pd.concat
df1

Series_Title Released_Year Runtime      Genre  \
2                        The Dark Knight          2008     152     Action   
21                          Interstellar          2014     169  Adventure   
23         Sen to Chihiro no kamikakushi          2001     125  Animation   
7                       Schindler's List          1993     195  Biography   
19                          Gisaengchung          2019     132     Comedy   
26                       La vita è bella          1997     116     Comedy   
1                          The Godfather          1972     175      Crime   
0               The Shawshank Redemption          1994     142      Drama   
688           E.T. the Extra-Terrestrial          1982     115     Family   
698  Willy Wonka & the Chocolate Factory          1971     100     Family   
321         Das Cabinet des Dr. Caligari          1920      76    Fantasy   
309                        The Third Man          1949     104  Film-Noir   
49                                Psycho          1960     109     Horror   
69                               Memento          2000     113    Mystery   
81                           Rear Window          1954     112    Mystery   
700                      Wait Until Dark          1967     108   Thriller   
12       Il buono, il brutto, il cattivo          1966     161    Western   

     IMDB_Rating              Director                Star1 No_of_Votes  \
2            9.0     Christopher Nolan       Christian Bale     2303232   
21           8.6     Christopher Nolan  Matthew McConaughey     1512360   
23           8.6        Hayao Miyazaki        Daveigh Chase      651376   
7            8.9      Steven Spielberg          Liam Neeson     1213505   
19           8.6          Bong Joon Ho         Kang-ho Song      552778   
26           8.6       Roberto Benigni      Roberto Benigni      623629   
1            9.2  Francis Ford Coppola        Marlon Brando     1620367   
0            9.3        Frank Darabont          Tim Robbins     2343110   
688          7.8      Steven Spielberg         Henry Thomas      372490   
698          7.8            Mel Stuart          Gene Wilder      178731   
321          8.1          Robert Wiene        Werner Krauss       57428   
309          8.1            Carol Reed         Orson Welles      158731   
49           8.5      Alfred Hitchcock      Anthony Perkins      604211   
69           8.4     Christopher Nolan           Guy Pearce     1125712   
81           8.4      Alfred Hitchcock        James Stewart      444074   
700          7.8         Terence Young       Audrey Hepburn       27733   
12           8.8          Sergio Leone       Clint Eastwood      688390   

           Gross  Metascore  
2    534858444.0       84.0  
21   188020017.0       74.0  
23    10055859.0       96.0  
7     96898818.0       94.0  
19    53367844.0       96.0  
26    57598247.0       59.0  
1    134966411.0      100.0  
0     28341469.0       80.0  
688  435110554.0       91.0  
698    4000000.0       67.0  
321  337574718.0        NaN  
309     449191.0       97.0  
49    32000000.0       97.0  
69    25544867.0       80.0  
81    36764313.0      100.0  
700   17550741.0       81.0  
12     6100000.0       90.0

In [41]:
def top_rated(df):
  df['rank']=df['IMDB_Rating'].rank(ascending=False)
  return df[df['rank']==1]
genres.apply(top_rated)

Series_Title Released_Year  Runtime  \
Genre                                                                   
Action    2                    The Dark Knight          2008      152   
Adventure 21                      Interstellar          2014      169   
Animation 23     Sen to Chihiro no kamikakushi          2001      125   
Biography 7                   Schindler's List          1993      195   
Crime     1                      The Godfather          1972      175   
Drama     0           The Shawshank Redemption          1994      142   
Fantasy   321     Das Cabinet des Dr. Caligari          1920       76   
Film-Noir 309                    The Third Man          1949      104   
Horror    49                            Psycho          1960      109   
Thriller  700                  Wait Until Dark          1967      108   
Western   12   Il buono, il brutto, il cattivo          1966      161   

                   Genre  IMDB_Rating              Director  \
Genre                                                         
Action    2       Action          9.0     Christopher Nolan   
Adventure 21   Adventure          8.6     Christopher Nolan   
Animation 23   Animation          8.6        Hayao Miyazaki   
Biography 7    Biography          8.9      Steven Spielberg   
Crime     1        Crime          9.2  Francis Ford Coppola   
Drama     0        Drama          9.3        Frank Darabont   
Fantasy   321    Fantasy          8.1          Robert Wiene   
Film-Noir 309  Film-Noir          8.1            Carol Reed   
Horror    49      Horror          8.5      Alfred Hitchcock   
Thriller  700   Thriller          7.8         Terence Young   
Western   12     Western          8.8          Sergio Leone   

                             Star1  No_of_Votes        Gross  Metascore  rank  
Genre                                                                          
Action    2         Christian Bale      2303232  534858444.0       84.0   1.0  
Adventure 21   Matthew McConaughey      1512360  188020017.0       74.0   1.0  
Animation 23         Daveigh Chase       651376   10055859.0       96.0   1.0  
Biography 7            Liam Neeson      1213505   96898818.0       94.0   1.0  
Crime     1          Marlon Brando      1620367  134966411.0      100.0   1.0  
Drama     0            Tim Robbins      2343110   28341469.0       80.0   1.0  
Fantasy   321        Werner Krauss        57428  337574718.0        NaN   1.0  
Film-Noir 309         Orson Welles       158731     449191.0       97.0   1.0  
Horror    49       Anthony Perkins       604211   32000000.0       97.0   1.0  
Thriller  700       Audrey Hepburn        27733   17550741.0       81.0   1.0  
Western   12        Clint Eastwood       688390    6100000.0       90.0   1.0

In [42]:
genres=movies.groupby('Genre')

In [43]:
## top rated movies in each genres
a=pd.DataFrame(columns=movies.columns)
for groups,data in genres:
  data['rank']=data['IMDB_Rating'].rank(ascending=False)
  a=pd.concat((a,data.head(1)))
a

Series_Title Released_Year Runtime      Genre  \
2                    The Dark Knight          2008     152     Action   
21                      Interstellar          2014     169  Adventure   
23     Sen to Chihiro no kamikakushi          2001     125  Animation   
7                   Schindler's List          1993     195  Biography   
19                      Gisaengchung          2019     132     Comedy   
1                      The Godfather          1972     175      Crime   
0           The Shawshank Redemption          1994     142      Drama   
688       E.T. the Extra-Terrestrial          1982     115     Family   
321     Das Cabinet des Dr. Caligari          1920      76    Fantasy   
309                    The Third Man          1949     104  Film-Noir   
49                            Psycho          1960     109     Horror   
69                           Memento          2000     113    Mystery   
700                  Wait Until Dark          1967     108   Thriller   
12   Il buono, il brutto, il cattivo          1966     161    Western   

     IMDB_Rating              Director                Star1 No_of_Votes  \
2            9.0     Christopher Nolan       Christian Bale     2303232   
21           8.6     Christopher Nolan  Matthew McConaughey     1512360   
23           8.6        Hayao Miyazaki        Daveigh Chase      651376   
7            8.9      Steven Spielberg          Liam Neeson     1213505   
19           8.6          Bong Joon Ho         Kang-ho Song      552778   
1            9.2  Francis Ford Coppola        Marlon Brando     1620367   
0            9.3        Frank Darabont          Tim Robbins     2343110   
688          7.8      Steven Spielberg         Henry Thomas      372490   
321          8.1          Robert Wiene        Werner Krauss       57428   
309          8.1            Carol Reed         Orson Welles      158731   
49           8.5      Alfred Hitchcock      Anthony Perkins      604211   
69           8.4     Christopher Nolan           Guy Pearce     1125712   
700          7.8         Terence Young       Audrey Hepburn       27733   
12           8.8          Sergio Leone       Clint Eastwood      688390   

           Gross  Metascore  rank  
2    534858444.0       84.0   1.0  
21   188020017.0       74.0   1.0  
23    10055859.0       96.0   1.0  
7     96898818.0       94.0   1.0  
19    53367844.0       96.0   1.5  
1    134966411.0      100.0   1.0  
0     28341469.0       80.0   1.0  
688  435110554.0       91.0   1.5  
321  337574718.0        NaN   1.0  
309     449191.0       97.0   1.0  
49    32000000.0       97.0   1.0  
69    25544867.0       80.0   1.5  
700   17550741.0       81.0   1.0  
12     6100000.0       90.0   1.0

In [44]:
movies['rankk'] = movies['IMDB_Rating'].rank(ascending=False)
movies

Series_Title Released_Year  Runtime   Genre  IMDB_Rating  \
0    The Shawshank Redemption          1994      142   Drama          9.3   
1               The Godfather          1972      175   Crime          9.2   
2             The Dark Knight          2008      152  Action          9.0   
3      The Godfather: Part II          1974      202   Crime          9.0   
4                12 Angry Men          1957       96   Crime          9.0   
..                        ...           ...      ...     ...          ...   
995    Breakfast at Tiffany's          1961      115  Comedy          7.6   
996                     Giant          1956      201   Drama          7.6   
997     From Here to Eternity          1953      118   Drama          7.6   
998                  Lifeboat          1944       97   Drama          7.6   
999              The 39 Steps          1935       86   Crime          7.6   

                 Director              Star1  No_of_Votes        Gross  \
0          Frank Darabont        Tim Robbins      2343110   28341469.0   
1    Francis Ford Coppola      Marlon Brando      1620367  134966411.0   
2       Christopher Nolan     Christian Bale      2303232  534858444.0   
3    Francis Ford Coppola          Al Pacino      1129952   57300000.0   
4            Sidney Lumet        Henry Fonda       689845    4360000.0   
..                    ...                ...          ...          ...   
995         Blake Edwards     Audrey Hepburn       166544  679874270.0   
996        George Stevens   Elizabeth Taylor        34075  195217415.0   
997        Fred Zinnemann     Burt Lancaster        43374   30500000.0   
998      Alfred Hitchcock  Tallulah Bankhead        26471  852142728.0   
999      Alfred Hitchcock       Robert Donat        51853  302787539.0   

     Metascore  rankk  
0         80.0    1.0  
1        100.0    2.0  
2         84.0    4.0  
3         90.0    4.0  
4         96.0    4.0  
..         ...    ...  
995       76.0  939.0  
996       84.0  939.0  
997       85.0  939.0  
998       78.0  939.0  
999       93.0  939.0  

[1000 rows x 11 columns]

In [45]:
# Functions  Groups which was just a mere string in loop here in function it act as literal for subgroup of partitioned dataframe
def rankk(group):
  group['rankk']=group['IMDB_Rating'].rank(ascending=False)
  return group.head(1)
genres.apply(rankk)

Series_Title Released_Year  Runtime  \
Genre                                                                   
Action    2                    The Dark Knight          2008      152   
Adventure 21                      Interstellar          2014      169   
Animation 23     Sen to Chihiro no kamikakushi          2001      125   
Biography 7                   Schindler's List          1993      195   
Comedy    19                      Gisaengchung          2019      132   
Crime     1                      The Godfather          1972      175   
Drama     0           The Shawshank Redemption          1994      142   
Family    688       E.T. the Extra-Terrestrial          1982      115   
Fantasy   321     Das Cabinet des Dr. Caligari          1920       76   
Film-Noir 309                    The Third Man          1949      104   
Horror    49                            Psycho          1960      109   
Mystery   69                           Memento          2000      113   
Thriller  700                  Wait Until Dark          1967      108   
Western   12   Il buono, il brutto, il cattivo          1966      161   

                   Genre  IMDB_Rating              Director  \
Genre                                                         
Action    2       Action          9.0     Christopher Nolan   
Adventure 21   Adventure          8.6     Christopher Nolan   
Animation 23   Animation          8.6        Hayao Miyazaki   
Biography 7    Biography          8.9      Steven Spielberg   
Comedy    19      Comedy          8.6          Bong Joon Ho   
Crime     1        Crime          9.2  Francis Ford Coppola   
Drama     0        Drama          9.3        Frank Darabont   
Family    688     Family          7.8      Steven Spielberg   
Fantasy   321    Fantasy          8.1          Robert Wiene   
Film-Noir 309  Film-Noir          8.1            Carol Reed   
Horror    49      Horror          8.5      Alfred Hitchcock   
Mystery   69     Mystery          8.4     Christopher Nolan   
Thriller  700   Thriller          7.8         Terence Young   
Western   12     Western          8.8          Sergio Leone   

                             Star1  No_of_Votes        Gross  Metascore  rankk  
Genre                                                                           
Action    2         Christian Bale      2303232  534858444.0       84.0    1.0  
Adventure 21   Matthew McConaughey      1512360  188020017.0       74.0    1.0  
Animation 23         Daveigh Chase       651376   10055859.0       96.0    1.0  
Biography 7            Liam Neeson      1213505   96898818.0       94.0    1.0  
Comedy    19          Kang-ho Song       552778   53367844.0       96.0    1.5  
Crime     1          Marlon Brando      1620367  134966411.0      100.0    1.0  
Drama     0            Tim Robbins      2343110   28341469.0       80.0    1.0  
Family    688         Henry Thomas       372490  435110554.0       91.0    1.5  
Fantasy   321        Werner Krauss        57428  337574718.0        NaN    1.0  
Film-Noir 309         Orson Welles       158731     449191.0       97.0    1.0  
Horror    49       Anthony Perkins       604211   32000000.0       97.0    1.0  
Mystery   69            Guy Pearce      1125712   25544867.0       80.0    1.5  
Thriller  700       Audrey Hepburn        27733   17550741.0       81.0    1.0  
Western   12        Clint Eastwood       688390    6100000.0       90.0    1.0

In [46]:
movies['IMDB_Rating'][movies['Genre']=='Action'].sum()

1367.3000000000002

In [47]:
# Functions
def rankk(group):
  return group['IMDB_Rating'].sum()
genres.apply(rankk)

Genre
Action       1367.3
Adventure     571.5
Animation     650.3
Biography     698.6
Comedy       1224.7
Crime         857.8
Drama        2299.7
Family         15.6
Fantasy        16.0
Film-Noir      23.9
Horror         87.0
Mystery        95.7
Thriller        7.8
Western        33.4
dtype: float64

In [48]:
# find ranking of each movie in the group according to IMDB score
def rank_movie(group):
  group['genre_rank'] = group['IMDB_Rating'].rank(ascending=False)
  return group

In [49]:
genres.apply(rank_movie)

Series_Title Released_Year  \
Genre                                                                           
Action   2                                      The Dark Knight          2008   
         5        The Lord of the Rings: The Return of the King          2003   
         8                                            Inception          2010   
         10   The Lord of the Rings: The Fellowship of the Ring          2001   
         13               The Lord of the Rings: The Two Towers          2002   
...                                                         ...           ...   
Thriller 700                                    Wait Until Dark          1967   
Western  12                     Il buono, il brutto, il cattivo          1966   
         48                        Once Upon a Time in the West          1968   
         115                         Per qualche dollaro in più          1965   
         691                             The Outlaw Josey Wales          1976   

              Runtime     Genre  IMDB_Rating           Director  \
Genre                                                             
Action   2        152    Action          9.0  Christopher Nolan   
         5        201    Action          8.9      Peter Jackson   
         8        148    Action          8.8  Christopher Nolan   
         10       178    Action          8.8      Peter Jackson   
         13       179    Action          8.7      Peter Jackson   
...               ...       ...          ...                ...   
Thriller 700      108  Thriller          7.8      Terence Young   
Western  12       161   Western          8.8       Sergio Leone   
         48       165   Western          8.5       Sergio Leone   
         115      132   Western          8.3       Sergio Leone   
         691      135   Western          7.8     Clint Eastwood   

                          Star1  No_of_Votes        Gross  Metascore  rankk  \
Genre                                                                         
Action   2       Christian Bale      2303232  534858444.0       84.0    4.0   
         5          Elijah Wood      1642758  377845905.0       94.0    7.0   
         8    Leonardo DiCaprio      2067042  292576195.0       74.0   11.0   
         10         Elijah Wood      1661481  315544750.0       92.0   11.0   
         13         Elijah Wood      1485555  342551365.0       87.0   16.0   
...                         ...          ...          ...        ...    ...   
Thriller 700     Audrey Hepburn        27733   17550741.0       81.0  645.0   
Western  12      Clint Eastwood       688390    6100000.0       90.0   11.0   
         48         Henry Fonda       302844    5321508.0       80.0   43.5   
         115     Clint Eastwood       232772   15000000.0       74.0  106.5   
         691     Clint Eastwood        65659   31800000.0       69.0  645.0   

              genre_rank  
Genre                     
Action   2           1.0  
         5           2.0  
         8           3.5  
         10          3.5  
         13          6.0  
...                  ...  
Thriller 700         1.0  
Western  12          1.0  
         48          2.0  
         115         3.0  
         691         4.0  

[1000 rows x 12 columns]

In [50]:
# find normalized IMDB rating group wise
def norm(group):
  group['norm_rating']=(group['IMDB_Rating']-group['IMDB_Rating'].min())/(group['IMDB_Rating'].max()-group['IMDB_Rating'].max())
  return group
movies.groupby('Genre',group_keys=True).apply(norm) # keys = True

Series_Title Released_Year  \
Genre                                                                           
Action   2                                      The Dark Knight          2008   
         5        The Lord of the Rings: The Return of the King          2003   
         8                                            Inception          2010   
         10   The Lord of the Rings: The Fellowship of the Ring          2001   
         13               The Lord of the Rings: The Two Towers          2002   
...                                                         ...           ...   
Thriller 700                                    Wait Until Dark          1967   
Western  12                     Il buono, il brutto, il cattivo          1966   
         48                        Once Upon a Time in the West          1968   
         115                         Per qualche dollaro in più          1965   
         691                             The Outlaw Josey Wales          1976   

              Runtime     Genre  IMDB_Rating           Director  \
Genre                                                             
Action   2        152    Action          9.0  Christopher Nolan   
         5        201    Action          8.9      Peter Jackson   
         8        148    Action          8.8  Christopher Nolan   
         10       178    Action          8.8      Peter Jackson   
         13       179    Action          8.7      Peter Jackson   
...               ...       ...          ...                ...   
Thriller 700      108  Thriller          7.8      Terence Young   
Western  12       161   Western          8.8       Sergio Leone   
         48       165   Western          8.5       Sergio Leone   
         115      132   Western          8.3       Sergio Leone   
         691      135   Western          7.8     Clint Eastwood   

                          Star1  No_of_Votes        Gross  Metascore  rankk  \
Genre                                                                         
Action   2       Christian Bale      2303232  534858444.0       84.0    4.0   
         5          Elijah Wood      1642758  377845905.0       94.0    7.0   
         8    Leonardo DiCaprio      2067042  292576195.0       74.0   11.0   
         10         Elijah Wood      1661481  315544750.0       92.0   11.0   
         13         Elijah Wood      1485555  342551365.0       87.0   16.0   
...                         ...          ...          ...        ...    ...   
Thriller 700     Audrey Hepburn        27733   17550741.0       81.0  645.0   
Western  12      Clint Eastwood       688390    6100000.0       90.0   11.0   
         48         Henry Fonda       302844    5321508.0       80.0   43.5   
         115     Clint Eastwood       232772   15000000.0       74.0  106.5   
         691     Clint Eastwood        65659   31800000.0       69.0  645.0   

              norm_rating  
Genre                      
Action   2            inf  
         5            inf  
         8            inf  
         10           inf  
         13           inf  
...                   ...  
Thriller 700          NaN  
Western  12           inf  
         48           inf  
         115          inf  
         691          NaN  

[1000 rows x 12 columns]

In [51]:
# groupby on multiple cols
duo = movies.groupby(['Director','Star1'])
duo
# size
duo.size()
# get_group
duo.get_group(('Aamir Khan','Amole Gupte'))   ############## (()) inputs

Series_Title Released_Year  Runtime  Genre  IMDB_Rating    Director  \
65  Taare Zameen Par          2007      165  Drama          8.4  Aamir Khan   

          Star1  No_of_Votes      Gross  Metascore  rankk  
65  Amole Gupte       168895  1223869.0        NaN   69.0

In [52]:
# find the most earning actor->director combo
duo['Gross'].sum().sort_values(ascending=False).head(1)

Director        Star1         
Akira Kurosawa  Toshirô Mifune    2.999877e+09
Name: Gross, dtype: float64

In [53]:
# find the best(in-terms of metascore(avg)) actor->genre combo
movies.groupby(['Star1','Genre'])['Metascore'].mean().reset_index().sort_values('Metascore',ascending=False).head(1)

Star1  Genre  Metascore
230  Ellar Coltrane  Drama      100.0

In [54]:
# agg on multiple groupby
# duo.agg(['min','max','mean'])

#Excercise

In [55]:
ipl = pd.read_csv('/content/Python/MyDrive/Colab Notebooks/Colab_Files_Data/Python/DataFrames/deliveries.csv')
ipl.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  ...  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0  ...         0   
1     2  DA Warner    S Dhawan  TS Mills              0  ...         0   
2     3  DA Warner    S Dhawan  TS Mills              0  ...         0   
3     4  DA Warner    S Dhawan  TS Mills              0  ...         0   
4     5  DA Warner    S Dhawan  TS Mills              0  ...         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs  player_dismissed dismissal_kind fielder  
0           0               NaN            NaN     NaN  
1           0               NaN            NaN     NaN  
2           4               NaN            NaN     NaN  
3           0               NaN            NaN     NaN  
4           2               NaN            NaN     NaN  

[5 rows x 21 columns]

In [56]:
ipl.shape

(179078, 21)

In [57]:
# find the top 10 batsman in terms of runs
ipl.groupby('batsman')['batsman_runs'].sum().sort_values(ascending=False).head(10)

batsman
V Kohli           5434
SK Raina          5415
RG Sharma         4914
DA Warner         4741
S Dhawan          4632
CH Gayle          4560
MS Dhoni          4477
RV Uthappa        4446
AB de Villiers    4428
G Gambhir         4223
Name: batsman_runs, dtype: int64

In [58]:
# find the batsman with max no of sixes
six = ipl[ipl['batsman_runs'] == 6]
six.groupby('batsman')['batsman'].count().sort_values(ascending=False).head(1).index[0]

'CH Gayle'

In [59]:
# find batsman with most number of 4's and 6's in last 5 overs
temp_df = ipl[ipl['over'] > 15]
temp_df = temp_df[(temp_df['batsman_runs'] == 4) | (temp_df['batsman_runs'] == 6)]
temp_df.groupby('batsman')['batsman'].count().sort_values(ascending=False).head(1).index[0]

'MS Dhoni'

In [60]:
temp_df=ipl[ipl['over'] >15]
temp_df=temp_df[(temp_df['batsman_runs']==4) | (temp_df['batsman_runs']==6)]
[temp_df['batsman'].value_counts()]

[batsman
 MS Dhoni          340
 AB de Villiers    208
 RG Sharma         208
 KA Pollard        206
 V Kohli           160
                  ... 
 DR Martyn           1
 JDS Neesham         1
 M Muralitharan      1
 GR Napier           1
 TA Boult            1
 Name: count, Length: 360, dtype: int64]

In [61]:
# find V Kohli's record against all teams
temp_df = ipl[ipl['batsman'] == 'V Kohli']
temp_df.groupby('bowling_team')['batsman_runs'].sum().reset_index()

bowling_team  batsman_runs
0       Chennai Super Kings           749
1           Deccan Chargers           306
2            Delhi Capitals            66
3          Delhi Daredevils           763
4             Gujarat Lions           283
5           Kings XI Punjab           636
6      Kochi Tuskers Kerala            50
7     Kolkata Knight Riders           675
8            Mumbai Indians           628
9             Pune Warriors           128
10         Rajasthan Royals           370
11   Rising Pune Supergiant            83
12  Rising Pune Supergiants           188
13      Sunrisers Hyderabad           509

In [62]:
# highest score of any batsman
ipl.groupby(['match_id','batsman'])['batsman_runs'].sum().sort_values(ascending=False)

match_id  batsman       
411       CH Gayle          175
60        BB McCullum       158
562       AB de Villiers    133
7935      RR Pant           130
620       AB de Villiers    129
                           ... 
455       DH Yagnik           0
27        V Kohli             0
119       MM Patel            0
          LRPL Taylor         0
203       FY Fazal            0
Name: batsman_runs, Length: 11294, dtype: int64

In [63]:
# Create a function that can return the highest score of any batsman
temp_df = ipl[ipl['batsman'] == 'V Kohli']
temp_df.groupby('match_id')['batsman_runs'].sum().sort_values(ascending=False).head(1).values[0]

113

In [64]:
def highest(batsman):
  temp_df = ipl[ipl['batsman'] == batsman]
  return temp_df.groupby('match_id')['batsman_runs'].sum().sort_values(ascending=False).head(1).values[0]
highest('DA Warner')

126

# Tasks

In [65]:
import pandas as pd
url = "https://docs.google.com/spreadsheets/d/e/2PACX-1vQjh5HzZ1N0SU7ME9ZQRzeVTaXaGsV97rU8R7eAcg53k27GTstJp9cRUOfr55go1GRRvTz1NwvyOnuh/pub?gid=1562145139&single=true&output=csv"
df = pd.read_csv(url)

# 1.   Using groupby make groups using the "Pclass" column and find out the average age and total number of missing values in the "Age" column for every group.

In [66]:
pclasses=df.groupby(['Pclass'])

In [67]:
# 1
avg_age=pclasses['Age'].mean().reset_index()

In [68]:
#2
temp_df=pd.DataFrame(df.columns)
for group,data in pclasses:
  print([group,data['Age'].isnull().sum()])

[(1,), 30]
[(2,), 11]
[(3,), 136]


# Q-2: Using groupby make groups using the "Pclass" column and fill every group's "Embarked" column's missing values with the mode value of that group. After that, print every group's "Embarked" column's value counts in ascending order.

In [69]:
print(df['Embarked'].value_counts(ascending=False).head(1).index[0])

S


In [70]:
pclasses=df.groupby('Pclass')
for group,data in pclasses:
  mode=data['Embarked'].value_counts(ascending=False).head(1).index[0]
  data=data['Embarked'].fillna(mode)

In [71]:
p=df.groupby('Pclass',group_keys=True)
def e(group):
  m=group['Embarked'].mode()
  group['Embarked']=group['Embarked'].fillna(m)
  return group
p.apply(e)

PassengerId  Survived  Pclass  \
Pclass                                      
1      1              2         1       1   
       3              4         1       1   
       6              7         0       1   
       11            12         1       1   
       23            24         1       1   
...                 ...       ...     ...   
3      882          883         0       3   
       884          885         0       3   
       885          886         0       3   
       888          889         0       3   
       890          891         0       3   

                                                         Name     Sex   Age  \
Pclass                                                                        
1      1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
       3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
       6                              McCarthy, Mr. Timothy J    male  54.0   
       11                            Bonnell, Miss. Elizabeth  female  58.0   
       23                        Sloper, Mr. William Thompson    male  28.0   
...                                                       ...     ...   ...   
3      882                       Dahlberg, Miss. Gerda Ulrika  female  22.0   
       884                             Sutehall, Mr. Henry Jr    male  25.0   
       885               Rice, Mrs. William (Margaret Norton)  female  39.0   
       888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
       890                                Dooley, Mr. Patrick    male  32.0   

            SibSp  Parch           Ticket     Fare Cabin Embarked  
Pclass                                                             
1      1        1      0         PC 17599  71.2833   C85        C  
       3        1      0           113803  53.1000  C123        S  
       6        0      0            17463  51.8625   E46        S  
       11       0      0           113783  26.5500  C103        S  
       23       0      0           113788  35.5000    A6        S  
...           ...    ...              ...      ...   ...      ...  
3      882      0      0             7552  10.5167   NaN        S  
       884      0      0  SOTON/OQ 392076   7.0500   NaN        S  
       885      0      5           382652  29.1250   NaN        Q  
       888      1      2       W./C. 6607  23.4500   NaN        S  
       890      0      0           370376   7.7500   NaN        Q  

[891 rows x 12 columns]

# 3. Q-3: Make groups based on "Embarked" column. And for each of this embarked group, make another group based on "Pclass" and find out the average fare (round off up to 2 decimal places) for each "Pclass" for each group of "Embarked".
Sample Output:

{'C': {1: 105, 2: 25, 3: 11},
 'Q': {1: 90, 2: 12, 3: 11},
 'S': {1: 70, 2: 20, 3: 15}}

In [72]:
emb=df.groupby(['Embarked'])
d={}
for group,data in emb:
  pclasses=data.groupby('Pclass')
  p=pclasses['Fare'].mean().round(2)      ## round ()
  d[group]=p.to_dict()                 ##  to_dict()
d

{('C',): {1: 104.72, 2: 25.36, 3: 11.21},
 ('Q',): {1: 90.0, 2: 12.35, 3: 11.18},
 ('S',): {1: 70.36, 2: 20.33, 3: 14.64}}

In [73]:
embarked_groups = df.groupby("Embarked")

# Questions Based on Fifa Worldcup - 2022 Dataset:
You can read the dataset by using the below sample code

import pandas as pd

In [74]:
import pandas as pd
df = pd.read_csv("https://docs.google.com/spreadsheets/d/e/2PACX-1vT3D_x_4DS6d51LKJ7ze1sxT5WpV5uiSVOFYHLwBiGru6vFyVv5h5-83AwFjxWYiWfCDjDAaarHAV-k/pub?gid=0&single=true&output=csv")

In [75]:
df

Sl. No  Match No.       Team      Against Group  Goal  Possession (%)  \
0         1          1      Qatar      Ecuador     A     0              40   
1         2          1    Ecuador        Qatar     A     2              46   
2         3          2    England         Iran     B     6              69   
3         4          2       Iran      England     B     2              20   
4         5          3    Senegal  Netherlands     A     0              39   
..      ...        ...        ...          ...   ...   ...             ...   
123     124         62    Morocco         Fram     F     0              55   
124     125         63    Croatia      Morocco     F     2              45   
125     126         63    Morocco      Croatia     F     1              45   
126     127         64  Argentina       France     C     3              46   
127     128         64     France    Argentina     D     3              40   

     Inside Penalty Area  Outside Penalty Area  Assists  ...  Fouls Against  \
0                      0                     0        0  ...             15   
1                      2                     0        1  ...             15   
2                      6                     0        6  ...              9   
3                      2                     0        1  ...             14   
4                      0                     0        0  ...             13   
..                   ...                   ...      ...  ...            ...   
123                    0                     0        0  ...             11   
124                    2                     0        2  ...             13   
125                    1                     0        0  ...             11   
126                    3                     0        1  ...             26   
127                    3                     0        1  ...             19   

     Offsides  Passes  Passes Completed  Crosses  Crosses Completed  Corners  \
0           3     453               387       10                  5        1   
1           4     484               419       26                 10        3   
2           2     810               733       29                  9        8   
3           2     232               156       11                  3        0   
4           2     391               326       22                  8        6   
..        ...     ...               ...      ...                ...      ...   
123         3     583               518       22                  1        3   
124         2     491               430       21                  3        6   
125         2     494               428       20                  5        3   
126         4     648               544       20                  4        6   
127         4     516               419       24                  3        5   

     Free Kicks  Penalties Scored  Pts  
0            19                 0    0  
1            17                 1    3  
2            16                 0    3  
3            10                 1    0  
4            14                 0    0  
..          ...               ...  ...  
123          15                 0    0  
124          13                 0    3  
125          15                 0    0  
126          22                 5    3  
127          28                 4    0  

[128 rows x 38 columns]

### `Q-4:` Perform `groupby` based on the `"Team"` column and then perform Z Normalization on top of the below columns of each group:
1. Passes
2. Passes Completed
3. Attempted Line Breaks
4. Completed Line Breaks

You have to make a python function named `z_normalization` which takes two arguments:

1. *group:* Every group that you have created
2. *cols_to_perform:* This parameter takes a list of columns on which you have to perform the Z-Normalization.

For this problem, you have to use th `apply()` method.

$\Large Z \ - \ Normalization = \frac{X_i - \mu}{std} $

After that find out the below values for each group:
- minimum "Passess"
- maximum "Passess"
- minimum "Yellow Cards"
- maximum "Yellow cards"
- average "Yellow Cards"
- maximum "Attempted Line Breaks"
- minimum "Attempted Line Breaks"
- standardard deviation of "Attempted Line Breaks"
- average Possession

In [76]:
df['Passes'].std()

160.1604871486747

In [77]:
team=df.groupby('Team',group_keys=False)

In [78]:
def norm(team):
  mean=team['Passes'].mean()
  std=team['Passes'].std()
  team['normalized_passes']=(team['Passes']-mean)/std
  return team
team.apply(norm)

Sl. No  Match No.       Team      Against Group  Goal  Possession (%)  \
0         1          1      Qatar      Ecuador     A     0              40   
1         2          1    Ecuador        Qatar     A     2              46   
2         3          2    England         Iran     B     6              69   
3         4          2       Iran      England     B     2              20   
4         5          3    Senegal  Netherlands     A     0              39   
..      ...        ...        ...          ...   ...   ...             ...   
123     124         62    Morocco         Fram     F     0              55   
124     125         63    Croatia      Morocco     F     2              45   
125     126         63    Morocco      Croatia     F     1              45   
126     127         64  Argentina       France     C     3              46   
127     128         64     France    Argentina     D     3              40   

     Inside Penalty Area  Outside Penalty Area  Assists  ...  Offsides  \
0                      0                     0        0  ...         3   
1                      2                     0        1  ...         4   
2                      6                     0        6  ...         2   
3                      2                     0        1  ...         2   
4                      0                     0        0  ...         2   
..                   ...                   ...      ...  ...       ...   
123                    0                     0        0  ...         3   
124                    2                     0        2  ...         2   
125                    1                     0        0  ...         2   
126                    3                     0        1  ...         4   
127                    3                     0        1  ...         4   

     Passes  Passes Completed  Crosses  Crosses Completed  Corners  \
0       453               387       10                  5        1   
1       484               419       26                 10        3   
2       810               733       29                  9        8   
3       232               156       11                  3        0   
4       391               326       22                  8        6   
..      ...               ...      ...                ...      ...   
123     583               518       22                  1        3   
124     491               430       21                  3        6   
125     494               428       20                  5        3   
126     648               544       20                  4        6   
127     516               419       24                  3        5   

     Free Kicks  Penalties Scored  Pts  normalized_passes  
0            19                 0    0           0.462725  
1            17                 1    3           1.085876  
2            16                 0    3           1.705844  
3            10                 1    0          -0.820695  
4            14                 0    0          -0.020934  
..          ...               ...  ...                ...  
123          15                 0    0           1.662218  
124          13                 0    3          -1.152198  
125          15                 0    0           0.896270  
126          22                 5    3           0.178979  
127          28                 4    0          -0.046643  

[128 rows x 39 columns]

In [79]:
teams=df.groupby('Team',group_keys=False)   ## Group keys False is very important
l=['Passes','Passes Completed','Attempted Line Breaks','Completed Line Breaks']
def z_normalization(group,l):
  n=l
  for i in n:
    group['normalized_{}'.format(i)]=(group[i]-group[i].mean())/(group[i].std())
  return (group)                                                ## return a group of each group sent and eventually gets appended in a single dataframe
df1=teams.apply(z_normalization,l)
df1

Sl. No  Match No.       Team      Against Group  Goal  Possession (%)  \
0         1          1      Qatar      Ecuador     A     0              40   
1         2          1    Ecuador        Qatar     A     2              46   
2         3          2    England         Iran     B     6              69   
3         4          2       Iran      England     B     2              20   
4         5          3    Senegal  Netherlands     A     0              39   
..      ...        ...        ...          ...   ...   ...             ...   
123     124         62    Morocco         Fram     F     0              55   
124     125         63    Croatia      Morocco     F     2              45   
125     126         63    Morocco      Croatia     F     1              45   
126     127         64  Argentina       France     C     3              46   
127     128         64     France    Argentina     D     3              40   

     Inside Penalty Area  Outside Penalty Area  Assists  ...  Crosses  \
0                      0                     0        0  ...       10   
1                      2                     0        1  ...       26   
2                      6                     0        6  ...       29   
3                      2                     0        1  ...       11   
4                      0                     0        0  ...       22   
..                   ...                   ...      ...  ...      ...   
123                    0                     0        0  ...       22   
124                    2                     0        2  ...       21   
125                    1                     0        0  ...       20   
126                    3                     0        1  ...       20   
127                    3                     0        1  ...       24   

     Crosses Completed  Corners  Free Kicks  Penalties Scored  Pts  \
0                    5        1          19                 0    0   
1                   10        3          17                 1    3   
2                    9        8          16                 0    3   
3                    3        0          10                 1    0   
4                    8        6          14                 0    0   
..                 ...      ...         ...               ...  ...   
123                  1        3          15                 0    0   
124                  3        6          13                 0    3   
125                  5        3          15                 0    0   
126                  4        6          22                 5    3   
127                  3        5          28                 4    0   

     normalized_Passes  normalized_Passes Completed  \
0             0.462725                     0.410365   
1             1.085876                     1.123095   
2             1.705844                     1.718012   
3            -0.820695                    -0.761575   
4            -0.020934                     0.032483   
..                 ...                          ...   
123           1.662218                     1.680153   
124          -1.152198                    -1.103970   
125           0.896270                     0.909216   
126           0.178979                    -0.028264   
127          -0.046643                    -0.280190   

     normalized_Attempted Line Breaks  normalized_Completed Line Breaks  
0                           -0.815368                         -1.153113  
1                           -0.039193                          0.817624  
2                           -1.172641                         -1.334904  
3                           -1.152297                         -1.147648  
4                           -1.077420                         -0.355371  
..                                ...                               ...  
123                         -0.496715                          0.899236  
124                         -0.555702                         -0.464907  
125                    

In [80]:
teams_agg=teams.agg({
    'Passes':['min','max'],
    'Yellow Cards':['min','max','mean'],
    'Attempted Line Breaks': ['min','max','std'],
    'Possession (%)':['mean']
})

# Questions on IPL wala dataset
ball by ball dataset - https://drive.google.com/file/d/1-kvv_9KCSAFWcrhS9WgTxSrURkRh6GNt/view?usp=share_link

In [93]:
balls = pd.read_csv("/content/Python/MyDrive/Colab Notebooks/Colab_Files_Data/Python/DataFrames/deliveries.csv")

In [82]:
def no(row):
  return row
balls.apply(no,axis=1)

match_id  inning         batting_team                 bowling_team  \
0              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
1              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
2              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
3              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
4              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
...          ...     ...                  ...                          ...   
179073     11415       2  Chennai Super Kings               Mumbai Indians   
179074     11415       2  Chennai Super Kings               Mumbai Indians   
179075     11415       2  Chennai Super Kings               Mumbai Indians   
179076     11415       2  Chennai Super Kings               Mumbai Indians   
179077     11415       2  Chennai Super Kings               Mumbai Indians   

        over  ball    batsman non_striker      bowler  is_super_over  ...  \
0          1     1  DA Warner    S Dhawan    TS Mills              0  ...   
1          1     2  DA Warner    S Dhawan    TS Mills              0  ...   
2          1     3  DA Warner    S Dhawan    TS Mills              0  ...   
3          1     4  DA Warner    S Dhawan    TS Mills              0  ...   
4          1     5  DA Warner    S Dhawan    TS Mills              0  ...   
...      ...   ...        ...         ...         ...            ...  ...   
179073    20     2  RA Jadeja   SR Watson  SL Malinga              0  ...   
179074    20     3  SR Watson   RA Jadeja  SL Malinga              0  ...   
179075    20     4  SR Watson   RA Jadeja  SL Malinga              0  ...   
179076    20     5  SN Thakur   RA Jadeja  SL Malinga              0  ...   
179077    20     6  SN Thakur   RA Jadeja  SL Malinga              0  ...   

        bye_runs  legbye_runs  noball_runs  penalty_runs  batsman_runs  \
0              0            0            0             0             0   
1              0            0            0             0             0   
2              0            0            0             0             4   
3              0            0            0             0             0   
4              0            0            0             0             0   
...          ...          ...          ...           ...           ...   
179073         0            0            0             0             1   
179074         0            0            0             0             2   
179075         0            0            0             0             1   
179076         0            0            0             0             2   
179077         0            0            0             0             0   

        extra_runs  total_runs  player_dismissed dismissal_kind    fielder  
0                0           0               NaN            NaN        NaN  
1                0           0               NaN            NaN        NaN  
2                0           4               NaN            NaN        NaN  
3                0           0               NaN            NaN        NaN  
4                2           2               NaN            NaN        NaN  
...            ...         ...               ...            ...        ...  
179073           0           1               NaN            NaN        NaN  
179074           0           2               NaN            NaN        NaN  
179075           0           1         SR Watson        run out  KH Pandya  
179076           0           2               NaN            NaN        NaN  
179077           0           0         SN Thakur            lbw        NaN  

[179078 rows x 21 columns]

# Q-5: Find batsman in below category-
1. Highest score(batsman) while chasing
2. Best Strike rate while chasing and have faced 100+ balls
Chasing mean team batting in second inning

In [83]:
#1
temp_df=balls[balls['inning']==2]
temp_df.groupby(['match_id','batsman'])['batsman_runs'].sum().sort_values(ascending=False).head(5)

match_id  batsman         
7953      SR Watson           121
243       PC Valthaty         120
279       V Sehwag            119
96        ST Jayasuriya       114
208       DPMD Jayawardene    110
Name: batsman_runs, dtype: int64

In [84]:
#
def ballss(balls):
  return (balls['over']-1) * 6 + ( balls['ball'])
balls['new_ball']=balls.apply(ballss, axis=1)

In [85]:
balls[(balls['match_id']==1) & (balls['batsman']=='CH Gayle') & (balls['wide_runs']==0)]['batsman_runs'].sum()

32

In [86]:
#2
temp_df=balls[(balls['inning']==2) & (balls['wide_runs']==0)]
teams=temp_df.groupby(['match_id','batsman'])
teams.groups.keys()
def strikerate(data):
  s=((data['batsman_runs'].sum())/(data['batsman_runs'].count())*100)
  return s
teams.apply(strikerate)

match_id  batsman      
1         A Choudhary      300.000000
          CH Gayle         152.380952
          KM Jadhav        193.750000
          Mandeep Singh    150.000000
          S Aravind          0.000000
                              ...    
11415     MS Dhoni          25.000000
          RA Jadeja        180.000000
          SK Raina          57.142857
          SN Thakur        100.000000
          SR Watson        137.288136
Length: 5521, dtype: float64

In [ ]:
bats=balls[(balls['inning']==2) & (balls['wide_runs']==0)].groupby('batsman')
def foo(batsman):
  if batsman['ball'].count() >100:
    return batsman['batsman_runs'].sum()/batsman['ball'].count()*100
  else:
    return 0
bats.apply(foo).sort_values(ascending=False)

### `Q-7`: Most successful batting pair in IPL. Batting pair who have scored most runs playing together.


In [97]:
def pair(df):
  return "-".join (df.sort_values().tolist())
balls['pair']=balls[['batsman','non_striker']].apply(pair,axis=1)

In [102]:
t=balls[['batsman','non_striker']].iloc[1000,:]
t.sort_values().tolist()

['KM Jadhav', 'SR Watson']

In [88]:
pairs=balls.groupby(['pair'])
pairs['batsman_runs'].sum().sort_values(ascending=False).head(10)

pair
AB de Villiers-V Kohli    2773
CH Gayle-V Kohli          2650
DA Warner-S Dhawan        2242
G Gambhir-RV Uthappa      1795
MS Dhoni-SK Raina         1411
M Vijay-MEK Hussey        1302
BB McCullum-DR Smith      1277
KA Pollard-RG Sharma      1251
MEK Hussey-SK Raina       1129
AT Rayudu-RG Sharma       1121
Name: batsman_runs, dtype: int64

# 8 Make a dataframe for all batting pairs played together.

In [89]:
balls.columns

Index(['match_id', 'inning', 'batting_team', 'bowling_team', 'over', 'ball',
       'batsman', 'non_striker', 'bowler', 'is_super_over', 'wide_runs',
       'bye_runs', 'legbye_runs', 'noball_runs', 'penalty_runs',
       'batsman_runs', 'extra_runs', 'total_runs', 'player_dismissed',
       'dismissal_kind', 'fielder', 'new_ball', 'pair'],
      dtype='object')

In [90]:
balls.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 179078 entries, 0 to 179077
Data columns (total 23 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   match_id          179078 non-null  int64 
 1   inning            179078 non-null  int64 
 2   batting_team      179078 non-null  object
 3   bowling_team      179078 non-null  object
 4   over              179078 non-null  int64 
 5   ball              179078 non-null  int64 
 6   batsman           179078 non-null  object
 7   non_striker       179078 non-null  object
 8   bowler            179078 non-null  object
 9   is_super_over     179078 non-null  int64 
 10  wide_runs         179078 non-null  int64 
 11  bye_runs          179078 non-null  int64 
 12  legbye_runs       179078 non-null  int64 
 13  noball_runs       179078 non-null  int64 
 14  penalty_runs      179078 non-null  int64 
 15  batsman_runs      179078 non-null  int64 
 16  extra_runs        179078 non-null  int

In [91]:
pd.set_option('display.max_columns',None)

In [92]:
temp_df = balls[balls['wide_runs']!=0]